example/mavlink_ftp_client/mavlink_ftp_client.cpp

```diff
@@ -0,0 +1,399 @@
+            std::cout << NORMAL_CONSOLE_TEXT << "\rUploading [" << std::setw(3) << percentage
+                      << "%] " << bytes_transferred << " of " << file_size;
+            if (bytes_transferred == file_size) {
+                std::cout << std::endl;
+            }
+        },
+        [prom](MavlinkFTP::Result result) { prom->set_value(result); });
+
+    return future_result.get();
+}
+
+MavlinkFTP::Result rename_file(
+    std::shared_ptr<MavlinkFTP>& mavlink_ftp,
+    const std::string& old_name,
+    const std::string& new_name)
+{
+    auto prom = std::make_shared<std::promise<MavlinkFTP::Result>>();
+    auto future_result = prom->get_future();
+    mavlink_ftp->rename_async(
+        old_name, new_name, [prom](MavlinkFTP::Result result) { prom->set_value(result); });
+
+    return future_result.get();
+}
+
+MavlinkFTP::Result get_file_checksum(
+    std::shared_ptr<MavlinkFTP>& mavlink_ftp, const std::string& remote_file_path, uint32_t& crc)
+{
+    auto prom = std::make_shared<std::promise<std::pair<MavlinkFTP::Result, uint32_t>>>();
+
+    auto future_result = prom->get_future();
+    mavlink_ftp->calc_file_crc32_async(
+        remote_file_path, [prom](MavlinkFTP::Result result, uint32_t checksum) {
+            prom->set_value(std::pair<MavlinkFTP::Result, uint32_t>(result, checksum));
+        });
+
+    std::pair<MavlinkFTP::Result, uint32_t> result = future_result.get();
+    if (result.first == MavlinkFTP::Result::SUCCESS) {
+        crc = result.second;
+    }
+    return result.first;
+}
+
+int main(int argc, char** argv)
+{
+    Mavsdk mavsdk;
+
+    auto prom = std::make_shared<std::promise<void>>();
+    auto future_result = prom->get_future();
+
+    std::cout << NORMAL_CONSOLE_TEXT << "Waiting to discover system..." << std::endl;
+    mavsdk.register_on_discover([prom](uint64_t uuid) {
+        std::cout << "Discovered system with UUID: " << uuid << std::endl;
+        prom->set_value();
+    });
+
+    std::string connection_url;
+    ConnectionResult connection_result;
+
+    if (argc >= 4) {
+        connection_url = argv[1];
+        connection_result = mavsdk.add_any_connection(connection_url);
+    } else {
+        usage(argv[0]);
+        return 1;
+    }
+
+    if (connection_result != ConnectionResult::SUCCESS) {
+        std::cout << ERROR_CONSOLE_TEXT
+                  << "Connection failed: " << connection_result_str(connection_result)
+                  << NORMAL_CONSOLE_TEXT << std::endl;
+        return 1;
+    }
+
+    future_result.get();
+
+    // We don't need to specify the UUID if it's only one system anyway.
+    // If there were multiple, we could specify it with:
+    // mavsdk.system(uint64_t uuid);
+    System& system = mavsdk.system();
+    auto mavlink_ftp = std::make_shared<MavlinkFTP>(system);
+    mavlink_ftp->set_timeout(200);
+    mavlink_ftp->set_retries(10);
+    mavlink_ftp->set_target_component_id(std::stoi(argv[2]));
```

Comment:
if `argv[2]` is not an int, it will call std::abort. I'd prefer a more graceful handling with proper error message.

Reply:
Changed.